src/hotspot/share/nmt/regionsTree.hpp

```diff
@@ -0,0 +1,171 @@
+  bool _with_storage;
+
+ public:
+  RegionsTree(bool with_storage) : VMATree() , _ncs_storage(with_storage), _with_storage(with_storage) { }
+
+  ReservedMemoryRegion find_reserved_region(address addr);
+
+  SummaryDiff commit_region(address addr, size_t size, const NativeCallStack& stack);
+  SummaryDiff uncommit_region(address addr, size_t size);
+
+  using Node = VMATree::TreapNode;
+
+  class NodeHelper {
+      Node* _node;
+      public:
+      NodeHelper() : _node(nullptr) { }
+      NodeHelper(Node* node) : _node(node) { }
+      inline bool is_valid() { return _node != nullptr; }
+      inline void clear_node() { _node = nullptr; }
+      inline VMATree::position position() { return _node->key(); }
+      inline bool is_committed_begin() { return ((uint8_t)out_state() & (uint8_t)VMATree::StateType::Committed) >= 2; }
+      inline bool is_released_begin() { return out_state() == VMATree::StateType::Released; }
+      inline bool is_reserved_begin() { return ((uint8_t)out_state() & (uint8_t)VMATree::StateType::Reserved) == 1; }
+      inline VMATree::StateType in_state() { return _node->val().in.type(); }
+      inline VMATree::StateType out_state() { return _node->val().out.type(); }
+      inline size_t distance_from(NodeHelper& other) { return position() - other.position(); }
+      inline NativeCallStackStorage::StackIndex out_stack_index() { return _node->val().out.stack(); }
+      inline MemTag in_tag() { return _node->val().in.mem_tag(); }
+      inline MemTag out_tag() { return _node->val().out.mem_tag(); }
+      inline void set_in_tag(MemTag tag) { _node->val().in.set_tag(tag); }
+      inline void set_out_tag(MemTag tag) { _node->val().out.set_tag(tag); }
+      inline void print_on(outputStream* st) {
+        auto st_str = [&](int s){
+          return s == (int)VMATree::StateType::Released ? "Rl" :
+                 s ==  (int)VMATree::StateType::Reserved ? "Rv" : "Cm";
+        };
+        st->print_cr("pos: " INTPTR_FORMAT " "
+                     "%s, %s <|> %s, %s",
+                     p2i((address)position()),
+                     st_str((int)in_state()),
+                     NMTUtil::tag_to_name(in_tag()),
+                     st_str((int)out_state()),
+                     NMTUtil::tag_to_name(out_tag())
+                     );
+      }
+    };
+
+  void print_on(outputStream* st) {
+    visit_in_order([&](Node* node) {
+      NodeHelper curr(node);
+      curr.print_on(st);
+      return true;
+    });
+  }
+
+  template<typename F>
+  void visit_committed_regions(position start, size_t size, F func) {
+    size_t end = start + size + 1;
+    size_t comm_size = 0;
+    size_t base = start;
+
+    NodeHelper prev;
+    visit_range_in_order(start, end, [&](Node* node) {
+      NodeHelper curr(node);
+      if (prev.is_valid()) {
+        base = prev.position();
+        if (prev.is_committed_begin()) {
+          comm_size += curr.distance_from(prev);
+          if (!curr.is_committed_begin()) {
```

Comment:
Ok I see. Thank you for clarifying.

Reply:
I have a couple more questions about this function:

Why bother keeping the variable `base`. It's updated whenever `prev` is updated, so why not just use `prev`?

Why do you allow for grouping two adjacent committed regions w/ diff `MemTags` together when creating `CommittedMemoryRegion`? Shouldn't they be treated as separate committed regions?  And if we are supposed to be grouping adjacent regions, shouldn't we be using the address of the first region, not the current one, when creating `CommittedMemoryRegion` (since we've been incrementing `comm_size`)?

Reply:
Your point on `base` is correct. `prev.position` can be used instead.

I agree that there is ambiguity in the method name and its args. This method SHOULD be called for a `ReservedMemoryRegion [base, base + size)`. This is implicitly assumed. The signature as `visit_committed_regions(ReservedMemoryRegion& rgn, F func)` is more declarative then.

With the above assumption, if two committed regions are adjacent with different `MemTag`, they certainly belong to different `ReservedMemoryRegion`s, one for CDS with `mtClassShared` and another for stack with `mtThreadStack` MemTag. This method doesn't care the `MemTag`s because all the sub-regions within the `ReservedMemoryRegion [base, base + size)` are assumed to have the same `MemTag`.

Reply:
Ok that makes sense to me if `visit_committed_regions` is only ever called on a single ReservedRegion.  But if that assumption is true, then isn't it also true that we should never encounter two adjacent committed regions (they should already be coalesced)? 

So then maybe you don't need the extra handling such as tracking `comm_size += curr.distance_from(prev)` or checking `if (!curr.is_committed_begin())`. Maybe you could just directly have something like:
```
if (prev.is_committed_begin()) {
    CommittedMemoryRegion cmr((address) prev.position(), curr.distance_from(prev), st);
...
}
```

Reply:
Very good comment. Thanks. Done.